File: src/cipher/AesCbc.zig
const std = @import("std");
const Data = @import("../Data.zig");
const AesEcb = @import("AesEcb.zig");

const Self = @This();

key: [16]u8,
iv: [16]u8,

pub fn decode(self: Self, data: *Data) !void {
    if (data.len() == 0) return;

    const allocator = data.allocator;
    const blocksize = 16;

    var xorTarget = try allocator.alloc(u8, data.len());
    defer allocator.free(xorTarget);
    @memcpy(xorTarget[0..blocksize], &self.iv);
    @memcpy(xorTarget[blocksize..], data.bytes[0..data.len() - blocksize]);

    const ecb = AesEcb{ .key = self.key, .pad = false };
    try data.decode(ecb);
    try data.xor(xorTarget);
    try data.unpad();
}

pub fn encode(self: Self, data: *Data) !void {
    if (data.len() == 0) return;

    const allocator = data.allocator;
    const blocksize = 16;

    try data.pad(blocksize);

    const ecb = AesEcb{ .key = self.key, .pad = false };

    var res = try allocator.alloc(u8, data.len());
    errdefer allocator.free(res);

    var prev: []const u8 = &self.iv;

    var windows = std.mem.window(u8, data.bytes, blocksize, blocksize);
    var i: u32 = 0;
    while (windows.next()) |block| {
        var dataBlock = try Data.copy(allocator, block);
        defer dataBlock.deinit();

        try dataBlock.xor(prev);
        try dataBlock.encode(ecb);
        @memcpy(res[i * blocksize..(i + 1) * blocksize], dataBlock.bytes);

        prev = block;
        i += 1;
    }

    data.reinit(res);
}

test "set 2 challenge 10" {
    const allocator = std.testing.allocator;

    const text = @embedFile("../data/10.txt");
    const size = std.mem.replacementSize(u8, text, "\n", "");
    const buf = try allocator.alloc(u8, size);
    defer allocator.free(buf);
    _ = std.mem.replace(u8, text, "\n", "", buf);

    var data = try Data.fromBase64(allocator, buf);
    defer data.deinit();

    var iv: [16]u8 = undefined;
    @memset(&iv, 0);

    const AES = Self{ .key = "YELLOW SUBMARINE".*, .iv = iv };
    try data.decode(AES);

    try std.testing.expectEqualStrings(@embedFile("../data/funky.txt"), data.bytes);
}

File: src/cipher/aes_cbc.rs
use aes::{
    Aes128,
    cipher::{BlockEncryptMut, KeyInit, generic_array::GenericArray},
};
use itertools::Itertools;

use crate::{
    Data, Error, Result,
    cipher::aes_ecb::AesEcb,
    error::{CipherError, CipherErrorType, InvalidLengthType},
};

use super::Cipher;

const BLOCKSIZE: u8 = 16;
const BLOCKSIZE_USIZE: usize = BLOCKSIZE as usize;
const CIPHER_NAME: &str = "AES-CBC";

/// Cipher for [AES-128] under [CBC mode].
///
/// [AES-128]: https://en.wikipedia.org/wiki/Advanced_Encryption_Standard
/// [CBC mode]: https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#CBC
#[derive(Debug, Clone)]
pub struct AesCbc {
    cipher: Aes128,
    iv: [u8; BLOCKSIZE_USIZE],
}

impl AesCbc {
    /// Create a new AES-CBC cipher with a specified key and initialization vector.
    pub fn new(key: impl AsRef<[u8]>, iv: impl AsRef<[u8]>) -> Result<Self> {
        let key = key.as_ref();
        let iv = iv.as_ref();

        let cipher = Aes128::new_from_slice(key).map_err(|_| {
            Error::CipherError(CipherError {
                cipher_name: CIPHER_NAME,
                kind: CipherErrorType::InvalidLength {
                    kind: InvalidLengthType::Key,
                    expected: BLOCKSIZE_USIZE,
                    actual: key.len(),
                },
            })
        })?;

        let iv = iv.try_into().map_err(|_| {
            Error::CipherError(CipherError {
                cipher_name: CIPHER_NAME,
                kind: CipherErrorType::InvalidLength {
                    kind: InvalidLengthType::IV,
                    expected: BLOCKSIZE_USIZE,
                    actual: iv.len(),
                },
            })
        })?;

        Ok(Self::init(cipher, iv))
    }

    /// Initialize an AES-CBC cipher with a pre-existing [`Aes128`] struct and an IV array.
    #[must_use]
    pub(crate) fn init(cipher: Aes128, iv: [u8; BLOCKSIZE_USIZE]) -> Self {
        Self { cipher, iv }
    }
}

impl Cipher for AesCbc {
    type Error = Error;

    fn decrypt(&mut self, data: &Data) -> Result<Data> {
        let mut ecb = AesEcb::init(self.cipher.clone(), false);
        let decoded = ecb.decrypt(data)?;

        let xor: Box<[u8]> = self
            .iv
            .iter()
            .chain(data.iter().dropping_back(BLOCKSIZE_USIZE))
            .copied()
            .collect();
        let xor = Data::from(xor);

        let res = decoded ^ xor;
        let res = res.unpad()?;
        Ok(res)
    }

    fn encrypt(&mut self, data: &Data) -> Result<Data> {
        let (_, bytes) = data.pad(BLOCKSIZE).chunks(BLOCKSIZE_USIZE).fold(
            (Data::from(self.iv), vec![]),
            |(prev, mut acc): (Data, Vec<u8>), data| {
                let data = Data::from(data);
                let mut xor = prev ^ &data;
                let bytes = GenericArray::from_mut_slice(&mut xor);
                self.cipher.encrypt_block_mut(bytes);
                acc.extend(bytes.as_slice());
                (data, acc)
            },
        );

        Ok(Data::from(bytes))
    }
}

#[cfg(test)]
mod tests {
    use miette::Result;
    use pretty_assertions::assert_eq;

    use super::*;

    #[test]
    fn s2c10_implement_cbc_mode() -> Result<()> {
        let text = include_str!("../../data/10.txt").replace('\n', "");
        let data = Data::from_base64(text)?;
        let mut cipher = AesCbc::new("YELLOW SUBMARINE", [0u8; BLOCKSIZE_USIZE])?;
        let res = cipher.decrypt(&data)?;

        assert_eq!(include_str!("../../data/funky.txt"), res.to_string());

        Ok(())
    }
}
